Lib/zipfile/__init__.py

```diff
@@ -1839,7 +1853,27 @@ def _sanitize_windows_name(cls, arcname, pathsep):
         arcname = pathsep.join(x for x in arcname if x)
         return arcname
 
-    def _extract_member(self, member, targetpath, pwd):
+    def _apply_permissions(self, member, path, mode):
+        """
+        Apply ZipFile permissions to a file on the filesystem with
+        specified PreserveMode
+        """
+        if mode == PreserveMode.NONE:
+            return
+
+        # Ignore permissions if the archive was created on Windows
+        if member.create_system == 0:
+            return
+
+        mask = {
+            PreserveMode.SAFE: 0o777,
+            PreserveMode.ALL: 0xFFFF,
+        }
+        new_mode = (member.external_attr >> 16) & mask[mode]
```

Comment:
I think this should honour the current umask. It's clumsy to get the current umask so the easiest way to do this is not to chmod the file after creation but to pass the requested mode bits into `os.open()` with `os.O_CREAT` when opening the target file for writing. It's also one fewer syscall so might be slightly faster.